File: app/src/main/java/com/fooddelivery/app/data/local/Converters.java
package com.fooddelivery.app.data.local;

import androidx.room.TypeConverter;
import com.fooddelivery.app.data.model.OrderItem;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Converters {
    
    @TypeConverter
    public String fromStringList(List<String> value) {
        if (value == null) return null;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < value.size(); i++) {
            sb.append(value.get(i));
            if (i < value.size() - 1) sb.append("|");
        }
        return sb.toString();
    }
    
    @TypeConverter
    public List<String> toStringList(String value) {
        if (value == null) return null;
        List<String> result = new ArrayList<>();
        String[] parts = value.split("\\|");
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty() && !trimmed.equals("null")) {
                result.add(trimmed);
            }
        }
        return result;
    }
    
    @TypeConverter
    public String fromDoubleList(List<Double> value) {
        if (value == null) return null;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < value.size(); i++) {
            sb.append(value.get(i));
            if (i < value.size() - 1) sb.append("|");
        }
        return sb.toString();
    }
    
    @TypeConverter
    public List<Double> toDoubleList(String value) {
        if (value == null) return null;
        List<Double> result = new ArrayList<>();
        String[] parts = value.split("\\|");
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty() && !trimmed.equals("null")) {
                try {
                    result.add(Double.parseDouble(trimmed));
                } catch (NumberFormatException e) {
                    // ignore
                }
            }
        }
        return result;
    }
    
    @TypeConverter
    public Long fromDate(Date value) {
        return value == null ? null : value.getTime();
    }
    
    @TypeConverter
    public Date toDate(Long value) {
        return value == null ? null : new Date(value);
    }
    
    @TypeConverter
    public String fromOrderItemList(List<OrderItem> value) {
        if (value == null) return null;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < value.size(); i++) {
            OrderItem item = value.get(i);
            sb.append(item.getDishId())
              .append("|").append(item.getName())
              .append("|").append(item.getPrice())
              .append("|").append(item.getQuantity())
              .append("|").append(item.getPackingFee());
            if (i < value.size() - 1) sb.append("@@");
        }
        return sb.toString();
    }
    
    @TypeConverter
    public List<OrderItem> toOrderItemList(String value) {
        if (value == null) return null;
        List<OrderItem> result = new ArrayList<>();
        String[] parts = value.split("@@");
        for (String part : parts) {
            if (!part.isEmpty()) {
                String[] data = part.split("\\|");
                if (data.length >= 5) {
                    try {
                        OrderItem item = new OrderItem(
                            Long.parseLong(data[0]),
                            data[1],
                            Double.parseDouble(data[2]),
                            Integer.parseInt(data[3]),
                            Double.parseDouble(data[4])
                        );
                        result.add(item);
                    } catch (NumberFormatException e) {
                        // ignore
                    }
                }
            }
        }
        return result;
    }
}

File: app/src/main/java/com/fooddelivery/app/data/local/Converters.kt
package com.fooddelivery.app.data.local

import androidx.room.TypeConverter
import com.fooddelivery.app.data.model.OrderItem
import java.util.Date

/**
 * 类型转换器 - 用于 Room 数据库处理复杂类型
 */
class Converters {
    
    @TypeConverter
    fun fromStringList(value: List<String>?): String? {
        return value?.joinToString(separator = "|")
    }
    
    @TypeConverter
    fun toStringList(value: String?): List<String>? {
        return value?.split("|")
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() && it != "null" }
    }
    
    @TypeConverter
    fun fromDoubleList(value: List<Double>?): String? {
        return value?.joinToString(separator = "|")
    }
    
    @TypeConverter
    fun toDoubleList(value: String?): List<Double>? {
        return value?.split("|")
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() && it != "null" }
            ?.map { it.toDouble() }
    }
    
    @TypeConverter
    fun fromDate(value: Date?): Long? {
        return value?.time
    }
    
    @TypeConverter
    fun toDate(value: Long?): Date? {
        return value?.let { Date(it) }
    }
    
    @TypeConverter
    fun fromOrderItemList(value: List<OrderItem>?): String? {
        return value?.joinToString(separator = "@@") { item ->
            "${item.dishId}|${item.name}|${item.price}|${item.quantity}|${item.packingFee}"
        }
    }
    
    @TypeConverter
    fun toOrderItemList(value: String?): List<OrderItem>? {
        return value?.split("@@")
            ?.filter { it.isNotEmpty() }
            ?.map { parts ->
                val data = parts.split("|")
                OrderItem(
                    dishId = data[0].toLong(),
                    name = data[1],
                    price = data[2].toDouble(),
                    quantity = data[3].toInt(),
                    packingFee = data[4].toDouble()
                )
            }
    }
}
